src/lib/adapters/waku/safe-waku.ts

```diff
@@ -0,0 +1,243 @@
+			: { startTime: endTime, endTime }
+		timeFilter = timeFilter || calculatedTimeFilter
+
+		const talkEmoji = isGroupChatId(chatId) ? '🗫' : '🗩'
+		this.log(`${talkEmoji}  subscribe to ${chatId}`)
+
+		const ws = makeWakustore(this.lightNode)
+		const unsubscribe = await ws.onSnapshot<Message>(
+			ws.collectionQuery('private-message', chatId, {
+				timeFilter,
+				pageDirection: PageDirection.FORWARD,
+				pageSize: 1000,
+			}),
+			(message) => this.queueMessage(callback, message, chatId),
+		)
+
+		const subscription = {
+			unsubscribe,
+			callback,
+		}
+
+		this.subscriptions.set(chatId, subscription)
+	}
+
+	async unsubscribeAll() {
+		for (const subscription of this.subscriptions.values()) {
+			await subscription.unsubscribe()
+		}
+
+		this.subscriptions = new Map()
+	}
+
+	async sendMessage(id: string, message: Message) {
+		if (!this.lightNode) {
+			this.lightNode = await connectWaku(this.options)
+		}
+
+		let error = undefined
+		let timeout = 1_000
+
+		const start = Date.now()
+
+		if (message.timestamp === this.lastSentTimestamp) {
+			message.timestamp++
+		}
+		this.lastSentTimestamp = message.timestamp
+
+		do {
+			try {
+				error = await sendMessage(this.lightNode, id, message)
+			} catch (e) {
+				error = e
+			} finally {
+				if (error) {
+					this.errors.numSendError++
+					this.log(`⁉️  Error: ${error}`)
+					this.log(`🕓 Waiting ${timeout} milliseconds...`)
+					await new Promise((r) => setTimeout(r, timeout))
+					if (timeout < 120_000) {
+						timeout += timeout
+					}
+				}
+			}
+		} while (error)
+
+		const elapsed = Date.now() - start
+
+		if (elapsed > 1000) {
+			this.log(`⏰ sending message took ${elapsed} milliseconds`)
+		}
+	}
+
+	private connectWaku() {
+		const waku = connectWaku({
+			onConnect: (connections) => {
+				this.log('✅ connected to waku', { connections })
+				this.safeResubscribe()
+
+				if (this.options?.onConnect) {
+					this.options?.onConnect(connections)
+				}
+			},
+			onDisconnect: () => {
+				this.log('❌ disconnected from waku')
+				this.errors.numDisconnect++
+
+				if (this.options?.onDisconnect) {
+					this.options.onDisconnect()
+				}
+			},
+		})
+		return waku
+	}
+
+	private async safeResubscribe() {
+		// eslint-disable-next-line no-constant-condition
+		while (true) {
+			try {
+				return await this.resubscribe()
+			} catch (e) {
+				console.error(`⁉️  ${e}`)
+			}
+		}
+	}
+
+	private async resubscribe() {
+		if (!this.lightNode) {
+			return
+		}
+
+		if (this.isSubscribing) {
+			return
+		}
+
+		this.isSubscribing = true
+
+		const chatIds = this.subscriptions.keys()
+		for (const subscription of this.subscriptions.values()) {
+			await subscription.unsubscribe()
+		}
+
+		const oldSubscriptions = this.subscriptions
+		this.subscriptions = new Map()
+
+		for (const chatId of chatIds) {
+			const callback = oldSubscriptions.get(chatId)?.callback
+			if (callback) {
+				await this.subscribe(chatId, undefined, callback)
+			}
+		}
```

Comment:
Maybe it would be good to `try`/`finally` this to avoid errors that could make it so that `this.isSubscribing = false` is never reached, even though some of these functions already have `try`/`catch`.

Reply:
On a second thought I am not sure the `isSubscribing` is really needed here. It would be better to make sure that the `safeSubscribe` is not called in parallel.

But you bring up a valid point that if the `subscribe` throws an exception currently it causes an unrecoverable error. I will see what can be done about it.